Create GeocodeEvictions_DC2024.sas

RAGarMir opened the pull request:
I have hit a roadblock. I am unable to get the Data Step (from step 2) to create a dataset the macro can read. The code I used for the Data Step starts after "**Creates new data set with new street_address and zip_char variables **;" I get the following message when running macro from Step 3, "ERROR: [DC_mar_geocode] The input data set {EVICTIONS_ADDRESSES} does not exist or could not be opened.
 the following error code when running the Macro from Step 3."

ptatian commented:
Hi @RAGarMir The curly braces ( { } ) in the sample code were meant to indicate that you should replace what is there with the appropriate names. But you should not keep the { }. If you do that, I believe your code will run properly.

RAGarMir commented:
@ptatian Looks like the program is running fine and I was also able to clean up all the addresses that did not match up. It should be ready for final review!!

RAGarMir commented:
> Thanks, @RAGarMir. I will review. But I need you to place a copy of the StreetAlt_GeoEvict_2024.txt file on SAS1, and your program should reference that copy. Otherwise, I will not be run the code.
**Where on SAS1 should I copy the file? Should it be the in the same Library and subdirectory location as in my local drive? What are the changes I should make to how the program is currently referenced to ensure the copy is also referenced?**
> And just to confirm, did you edit any of the addresses directly in the input data file (FY24DetailedData(records).csv)? Or are all the corrections done through the StreetAlt_GeoEvict_2024.txt file?
**I made some changes to the input data file as well. It was mostly spelling mistakes for places like Minnesota Ave.**

ptatian commented:
Actually, I'm sorry. StreetAlt_GeoEvict_2024.txt is fine where it is! 

Thank you.

ptatian reviewed (changes requested):
Hi @RAGarMir . Thanks for doing this. You did a good job of implementing the macro according to the instructions. I added to your code some things that you would not have known to do, but I included notes in the review to point out those changes so that you can pick up some new lessons from them. **Also, there are a few changes below that I'd like to ask you to make. **

_Street name mispellings_

It is preferrable to correct the street name mispellings in the data by using the Street_alt file, rather than editing the input data. For example, 
```
WARNING: [DC_mar_geocode] Street not found: HAWTHORONE ( street_address=86 Hawthorone Court #86 / _N_=768 )
```
Can be addressed by adding this line to Street_alt:
```
HAWTHORONE,HAWTHORNE
```

Same for WLIMINGTON, 14H, etc. SInce it is all working, you don't need to change anything. I just mentioning for future reference. 

_Address matching_

Other data entry problems do need to be addressed in the input data, like you have done. There is still one address that is not being read in properly because of a data entry error. 
```
WARNING: [DC_mar_geocode] Street not found: 19TH STREETSE ( street_address=1701 19th StreetSE #203 / _N_=1548 )
```
The quadrant SE got added to the street name. **So if you would please fix that, that would be great.** 

This address also shows up in the "UNMATCHED OR NONEXACT ADDRESSES" list in the output. Scanning this list, you can see that 46 addresses do not geocode to the same exact address. For example, "231 37th Street SE" was geocoded to "232 37th St SE." This could be because of data entry error (for example, perhaps it should have been 231 37th PLACE SE, which is a valid address), or it could be a problem in our address file. Since we have no way of knowing for sure, we will just have to leave these addresses as is. 

In any case, the approximate matches look like they are close to the entered one, so for things like summarizing data by wards or neighborhoods, we should be OK. If we want to match the data to parcel records, however, we will not want to use any of the non-exact matches, which would match to the wrong parcel. 

_Other data problems_

In the frequency tables of DISPOSITION, one row has CANCELED spelled CANCELLED. Since we want categorical variable values to be consistent, we should correct this. Again, **it is generally better to correct errors in the code rather than editing data files,** which is less transparent. In addition, this problem may appear in future data files, so handling it code will be valuable. **In the `data evictions_addresses;` data step, please add an IF/THEN statement to correct the mispelled value.**

We also have 6 rows with missing values for YEAR, and 1 with YEAR=2025. The year is  usually part of CASE_NUM, but for the 6 missing the CASE_NUM does not have a part that looks like a year. And for the 2025 one, the case number is 3278-25. So, not anything we can do about that. 

BTW, the data do show that almost all these evictions (>98%) are cases from 2022 or later. 

Please make the changes indicated above, and then tag me again on the pull request. Thank you! 

cc: @lhendey

RAGarMir commented on Prog/2024/GeocodeEvictions_DC2024.sas:
Added if/then statement to correct "cancelled" Spelling

RAGarMir replied:
@ptatian I am not sure why I does not work. After running the code and looking at the outputs I assume that the if/then statement is in the wrong place and the line with the "Cancelled" disposition is filtered out at an earlier step.

ptatian replied:
Condition statements in SAS are case-sensitive by default. If you write `disposition = 'cancelled'`, that is only true if the value is all lower case. But, in the data the disposition values are uppercase (CANCELLED). 

It's better, however, not to assume the case of the values. In a future data file they might be different. You can use the upcase() function to convert the values of disposition to all upper case for the comparison: `upcase(disposition) = 'CANCELLED'`.

And, then, you want to assign the new value in uppercase as well so that they are consistent. `'CANCELED'`

ptatian replied:
I reviewed the default StreetAlt file and found some other problem entries. I've fixed the on in the MAR library but for your StreetAlt_GeoEvict_2024.txt file, you will want to delete these rows as well.

```
MC KENNA,MCKENNA
SURCUM CORDA,SURSUM CORDA
SURSUM CURDA,SURSUM CORDA
VANDENBURG,VANDENBERG
YONT,YOUNT
```

RAGarMir commented:
@ptatian I resolved the issues mentioned above. I did come across a new issue. I am getting the following error code:
ERROR: [StreetAlt] Invalid entry for correct spelling of street name LACKLAND WAY G BA  as LACKLAND

It looks like "LACKLAND" is not in the ValidStreets.html file. I am unsure whether I should add it to the file or find another way to resolve the error.

ptatian commented:
@RAGarMir Regarding Lackland, what the error message means is that Lackland was used as a correct street name spelling in StreetAlt, but there is no such street in DC. It may have been in the original StreetAlt file, and perhaps there was such a street earlier. 

The solution is simply to delete this row from Prog\2024\StreetAlt_GeoEvict_2024.txt
```
LACKLAND WAY G BA,LACKLAND
```

RAGarMir commented:
@ptatian I updated the StreetAlt_GeoEvict2024.txt file and removed all problem entries. 

I also updated the GeocodeEvictions_DC2024.sas file to include the upcase function within the added if/then statement. I ran the code without errors. I was not able to tell if the upcase function did fix the problem with the disposition='CANCELLED' entry.

RAGarMir commented:
@ptatian I could not find the CANCELLED entry in the output. I am assuming everything worked correctly.

ptatian commented:
That sounds good!

ptatian commented:
@RAGarMir This code is ready to go, but I'd like to move it to a new Evict repo that I have created. We will not merge this pull request in the Requests repo, but finish it up in the new repo.  

Please see https://github.com/NeighborhoodInfoDC/Evict/issues/1 for instructions.
